File: module-infra/src/main/java/maple/expectation/infrastructure/persistence/entity/CharacterEquipmentJpaEntity.java
package maple.expectation.infrastructure.persistence.entity;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import maple.expectation.util.converter.GzipStringConverter;

/**
 * JPA Entity for Character Equipment persistence.
 *
 * <p>This is a PERSISTENCE entity with JPA annotations. It belongs to the infrastructure layer and
 * should only be used by repository implementations.
 *
 * <p><b>Important:</b> Business logic has been moved to {@link
 * maple.expectation.domain.model.equipment.CharacterEquipment}. This entity is purely for database
 * mapping.
 *
 * @see maple.expectation.domain.model.equipment.CharacterEquipment
 */
@Entity
@Table(
    name = "character_equipment",
    indexes = @Index(name = "idx_character_equipment_updated_at", columnList = "updated_at"))
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class CharacterEquipmentJpaEntity {

  @Id
  @Column(length = 100)
  private String ocid;

  @Convert(converter = GzipStringConverter.class)
  @Lob
  @Column(columnDefinition = "LONGBLOB", nullable = false)
  private String jsonContent;

  private LocalDateTime updatedAt;

  /**
   * Creates a new CharacterEquipmentJpaEntity.
   *
   * @param ocid the character OCID (primary key)
   * @param jsonContent the equipment JSON content (GZIP compressed)
   * @return new CharacterEquipmentJpaEntity instance with current timestamp
   */
  @Builder
  public CharacterEquipmentJpaEntity(String ocid, String jsonContent) {
    this.ocid = ocid;
    this.jsonContent = jsonContent;
    this.updatedAt = LocalDateTime.now(); // Default: current time
  }

  /**
   * Creates a new CharacterEquipmentJpaEntity with explicit timestamp.
   *
   * <p>ADR-084: This constructor overload allows preserving the domain entity's updatedAt timestamp
   * when mapping to JPA, avoiding timestamp loss.
   *
   * @param ocid the character OCID (primary key)
   * @param jsonContent the equipment JSON content (GZIP compressed)
   * @param updatedAt the update timestamp (preserved from domain entity)
   * @return new CharacterEquipmentJpaEntity instance
   */
  public static CharacterEquipmentJpaEntity of(
      String ocid, String jsonContent, LocalDateTime updatedAt) {
    CharacterEquipmentJpaEntity entity = new CharacterEquipmentJpaEntity();
    entity.ocid = ocid;
    entity.jsonContent = jsonContent;
    entity.updatedAt = updatedAt != null ? updatedAt : LocalDateTime.now();
    return entity;
  }

  /**
   * Updates the equipment data and timestamp.
   *
   * <p><b>Note:</b> This method exists for JPA/Hibernate usage. Domain-level updates should use the
   * domain entity's {@link maple.expectation.domain.model.equipment.CharacterEquipment#updateData
   * updateData()} method.
   *
   * @param newJsonContent the new JSON content
   */
  public void updateData(String newJsonContent) {
    this.jsonContent = newJsonContent;
    this.updatedAt = LocalDateTime.now();
  }
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/persistence/entity/CharacterEquipmentJpaEntity.kt
package maple.expectation.infrastructure.persistence.entity

import jakarta.persistence.*
import java.time.LocalDateTime
import maple.expectation.util.converter.GzipStringConverter
import org.hibernate.annotations.Index

/**
 * JPA Entity for Character Equipment persistence.
 *
 * <p>This is a PERSISTENCE entity with JPA annotations. It belongs to the infrastructure layer and
 * should only be used by repository implementations.
 *
 * <p><b>Important:</b> Business logic has been moved to
 * [maple.expectation.domain.model.equipment.CharacterEquipment]. This entity is purely for database
 * mapping.
 *
 * @see maple.expectation.domain.model.equipment.CharacterEquipment
 */
@Entity
@Table(indexes = [Index(name = "idx_character_equipment_updated_at", columnList = "updated_at")])
open class CharacterEquipmentJpaEntity {

    @Id
    @Column(length = 100)
    open var ocid: String? = null

    @Convert(converter = GzipStringConverter::class)
    @Column(columnDefinition = "BYTEA", nullable = false)
    open var jsonContent: String? = null

    open var updatedAt: LocalDateTime? = null

    /**
     * Creates a new CharacterEquipmentJpaEntity.
     *
     * @param ocid the character OCID (primary key)
     * @param jsonContent the equipment JSON content (GZIP compressed)
     */
    constructor(ocid: String?, jsonContent: String?) {
        this.ocid = ocid
        this.jsonContent = jsonContent
        this.updatedAt = LocalDateTime.now()
    }

    /**
     * Default constructor for JPA.
     */
    protected constructor()

    /**
     * Creates a new CharacterEquipmentJpaEntity with explicit timestamp.
     *
     * <p>ADR-084: This factory method allows preserving the domain entity's updatedAt timestamp
     * when mapping to JPA, avoiding timestamp loss.
     *
     * @param ocid the character OCID (primary key)
     * @param jsonContent the equipment JSON content (GZIP compressed)
     * @param updatedAt the update timestamp (preserved from domain entity)
     */
    companion object {
        @JvmStatic
        fun of(
            ocid: String?,
            jsonContent: String?,
            updatedAt: LocalDateTime?,
        ): CharacterEquipmentJpaEntity {
            val entity = CharacterEquipmentJpaEntity()
            entity.ocid = ocid
            entity.jsonContent = jsonContent
            entity.updatedAt = updatedAt ?: LocalDateTime.now()
            return entity
        }
    }

    /**
     * Updates the equipment data and timestamp.
     *
     * <p><b>Note:</b> This method exists for JPA/Hibernate usage. Domain-level updates should use the
     * domain entity's
     * [maple.expectation.domain.model.equipment.CharacterEquipment.updateData]
     * method.
     *
     * @param newJsonContent the new JSON content
     */
    open fun updateData(newJsonContent: String?) {
        this.jsonContent = newJsonContent
        this.updatedAt = LocalDateTime.now()
    }
}
